optimizing free_nodes() function in server.

bremanandjk opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature

On every invocation of free_nodes(), especially each job exit, it is consuming more time for releasing the nodes associated with the job object.

#### Describe Your Change

Optimized in a way that instead of looping through to "each subnode" in "each node" in "each mom", identified the exact node directly from chunks and releasing nodes using deallocate_job_from_node() function.


#### Link to Design Doc
<!--- If there is a design, link to it here: **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)** -->


#### Attach Test and Valgrind Logs/Output
<!--- Please attach your test log output from running the test you added (or from existing tests that cover your changes) -->
<!--- Don't forget to run Valgrind if appropriate and attach the resulting logs -->
Executed the below steps to identify the time consumption in each invocation. 

a) Install pbs
b) kill running mom kill -9 $(pidof pbs_mom)
c) run pbs_mom -m as root to start mock mom
d) pbs_config --vnodify -N 10 -a resources_available.ncpus=500 (to create 10 vnodes with 500 ncpus each)
e) turn off scheduling qmgr -c "s s scheduling=0"
f) submit 50k jobs (for i in $(seq 1 1 5000); do qsub -koe -- /bin/date; done; as pbsroot user)
g) start only one sched cycle (qmgr -c "s s scheduling=1" && qmgr -c "s s scheduling=0")
h) wait for jobs to finish

Before fix:
<img width="1662" alt="Before_fix_2_master" src="https://user-images.githubusercontent.com/17980660/88009680-14954280-cae1-11ea-9192-cab47fe89a51.png">

After fix:
<img width="1662" alt="After_fix_2_master" src="https://user-images.githubusercontent.com/17980660/88009693-19f28d00-cae1-11ea-8b9c-23e27ffb99d5.png">


Platforms: CENTOS7,UBUNTU1804,SUSE15
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|2708|1794|2|0|0|0|1792|


<!--- Pull Request Guidelines: [Pull Request Guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/1187348483/Pull+Request+Guidelines) -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/openpbs/openpbs?pullRequest=1914) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/openpbs/openpbs?pullRequest=1914) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/openpbs/openpbs?pullRequest=1914) it.</sub>

bremanandjk commented:
Thanks for your reviews, I will remove the loop which iterates over mominfo structures to disassociate jobs in it. Let me replace those functionalities by a performant code to get a high percentage results.

nithinj commented on src/server/node_manager.c:
```diff
@@ -6850,124 +6850,100 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
+	/* decrement number of jobs on the Mom who is the first Mom */	
+	/* for the job, Mother Superior; incremented in set_nodes() */	
+	/* and saved in ji_destin in assign_hosts()		    */	
+	if (((pjob->ji_qs.ji_svrflags & JOB_SVFLG_HasNodes) != 0) &&	
```
this can be part of deallocate_job_from_node instead of handling separately.

bremanandjk replied:
Yes, that is what I am trying now.

hirenvadalia replied:
@bremanandjk Looks like you missed push this change?

bremanandjk replied:
On debugging further, I have realized we should not add this code in deallocate_job_from_node(). So left this code as it is.

nithinj commented on src/server/node_manager.c:
```diff
@@ -6850,124 +6850,100 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
+
+		for (j=0; j<psvrmom->msr_jbinxsz; j++) {	
+			if (psvrmom->msr_jobindx[j] == pjob) {	
+				psvrmom->msr_jobindx[j] = NULL;	
```
Find the mom from node struct rather than looping through.

bremanandjk commented:
> I do not see any point of this change due to two things:
> 
> 1. We are still looping over every mom in the system (think of an exascale system with 100k nodes)
> 2. For a multivnoded system, the gain is very miniscule, from 2% to .5% - not seeing much justification. Maybe for a very large multivnoded system there is some ROI, but then not sure how common those are (and what is that typical size)?
> 
> In other words, the enhancement is not buying us enough. I would suggest rework

I have modified the code such that not to parse all the moms in the system. Now we are considering only the moms associated with the node object. Thereby, this change would give a dramatic performance gain. 

Also, I have modified the test to show the substantial results, test steps follows. 

Test setup:

1) Install pbs
2) kill running mom kill -9 $(pidof pbs_mom)
3) run pbs_mom -m as root to start mock mom
4) pbs_config --vnodify -N 100 -a resources_available.ncpus=500 (to create 100 vnodes with 500 ncpus each)
5) Run 50 moms with unique port.
6) turn off scheduling qmgr -c "s s scheduling=0"
7) submit 50k jobs (for i in $(seq 1 1 50000); do qsub -koe -- /bin/date; done; as pbsroot user)
8) start only one sched cycle (qmgr -c "s s scheduling=1" && qmgr -c "s s scheduling=0")
9) wait for jobs to finish
10) Run perf tools to gauge the performance. 

Before fix: 
<img width="1662" alt="trial_2_before_fix" src="https://user-images.githubusercontent.com/17980660/88133343-cd29b780-cbaf-11ea-8773-4a89efa85bd2.png">

After fix: 
<img width="1662" alt="trial_2_after_fix" src="https://user-images.githubusercontent.com/17980660/88133350-d581f280-cbaf-11ea-8c12-e5c40444d49c.png">

TH3 results:
<img width="1568" alt="Screen Shot 2020-07-22 at 12 05 56 AM" src="https://user-images.githubusercontent.com/17980660/88133365-e03c8780-cbaf-11ea-8718-4c59a2457e9b.png">

P.S: I have maintained the setup I used in the test machine as it is. If there are any issues in my test to set up, please let me know. I would redo the tests. Thanks

hirenvadalia commented on src/server/node_manager.c:
```diff
@@ -6839,6 +6839,51 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
 }
 
+void
+disassociate_job_from_moms(job *pjob, struct pbsnode *pnode)
```
I don't think we need this, we can put this code in `deallocate_job_from_node()`

hirenvadalia replied:
Thanks, @bremanandjk for the explanation. I got it that we can't put this in `deallocate_job_from_node()`. But I would suggest to make it static and name it like `remove_job_index_from_mom()` (as we have `add_job_index_in_mom()` in the same file, which is exactly reverse of this). Also please add function header.

bremanandjk replied:
I have renamed the function name.

hirenvadalia commented on src/server/node_manager.c:
```diff
@@ -6839,6 +6839,51 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
+			}	
+		}
+		
+		if (special_case) {	
```
Do we really need this now? I think we can get rid of this

bremanandjk replied:
This portion of code was already present for some other bug debugging, If the developer who added this code felt that it solved their purpose, let them remove it. That's why I preserved this change as it is. It is out of the scope of job disassociation from node and mom PR.

hirenvadalia replied:
This code was added ~3 years back and I think it served its purpose. But still, let's hear from the author of this code, @nithinj Do we need this special case of debug message anymore? If not then I would like to get rid of it.

bremanandjk replied:
As I had mentioned earlier, whether we decide to remove this bug specific debugging code or not, that should be handled separately by the original developer in separate PR. I don't want to delay this PR merge further with the "out of scope" code changes.

nithinj replied:
This is added at least two releases before. If it did not have helped so far, then I don't see any point in keeping it.

hirenvadalia commented on src/server/node_manager.c:
```diff
@@ -6839,6 +6839,51 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
+			continue;	
+		psvrmom = (mom_svrinfo_t *)(pnode->nd_moms[i]->mi_data);	
+
+		for (j=0; j<psvrmom->msr_jbinxsz; j++) {	
```
I think we should work on this and try to get rid of this loop also, as it is going to all jobs assigned to mom

bremanandjk replied:
In general, its good to remove the loop. But here we need to consider this case in detail. 
Whenever a job queued in MoM, the job object would be registered in mominfo structure and the same would be removed after completion. For short jobs, it is so quick.  Even if we try with AVL tree DS to store job id versus mominfo index, the operation would be more complex in adding/removing/ frequently. The gain we get on from that would be trivial. (We are already getting 93% to 0.5 %).  I feel those operations are effective in terms of a large number of elements like tens of hundreds of thousands. 

This in case, at any moment, the number of runtime jobs is close to a few hundred, I would prefer the current code is much better in comparison to having another DS. If you think otherwise, plz explain with a technical explanation, I will consider in another PR.

hirenvadalia replied:
Thanks, @bremanandjk for the explanation, but my comment was in a generic or architectural way (sorry if it was not clear enough), as currently our mom structures are using array and looping through those arrays, which is slow, we can refactor those and make them faster.

hirenvadalia commented on src/server/node_manager.c:
```diff
@@ -6850,124 +6895,69 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
+			return;
+	} else {
+		log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_JOB, LOG_DEBUG,
+		pjob->ji_qs.ji_jobid, "in free_nodes and no exec_vnode");
```
please merge this with the above line

bremanandjk replied:
Agree, I have addressed it.

hirenvadalia commented on src/server/node_manager.c:
```diff
@@ -6850,124 +6895,69 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
-		}
+	/*  remove this jobs's jobinfo entry from each vnode   */
+	
+	if (pjob->ji_wattr[(int)JOB_ATR_exec_vnode].at_flags & ATR_VFLAG_SET) {
```
Do we really need this if/else? old code was not checking for this, and I don't think free_nodes() will be called if job is allocated any nodes

bremanandjk replied:
It is a good practice to check the attribute flag and retrieve it. I am seeing this practice everywhere. please refer same file, you could see many references.

hirenvadalia replied:
I agree that its good practice checks things before accessing, but we can improve this if/else as below:
```c
if ((pjob->ji_wattr[JOB_ATR_exec_vnode].at_flags & ATR_VFLAG_SET) == 0 || ((execvnod_in = pjob->ji_wattr[JOB_ATR_exec_vnode].at_val.at_str) == NULL)) {
	log_event(PBSEVENT_DEBUG2, PBS_EVENTCLASS_JOB, LOG_DEBUG, pjob->ji_qs.ji_jobid, "in free_nodes and no exec_vnode");
	return;
}
```

bremanandjk replied:
I have changed this code.

agrawalravi90 commented on src/server/node_manager.c:
```diff
@@ -6850,124 +6895,69 @@ set_nodes(void *pobj, int objtype, char *execvnod_in, char **execvnod_out, char
-				if (pjob->ji_qs.ji_substate == JOB_SUBSTATE_PROVISION)
-					free_prov_vnode(pnode);
-			}
+	chunk = parse_plus_spec_r(execvncopy, &last, &hasprn);
```
I'm a bit concerned that parse_plus_spec can be slow for long exec_vnodes, can you please also do a performance comparison with multi-node jobs which occupy 100+ nodes and post results before and after your changes?

bremanandjk replied:
This parse_plus_spec_r() is used extensively in all many places in Server, Sched & MoM. While optimizing this function in general, this code also would get optimized. 

Let me try a multi-node job test also and post the results. Thanks for your review.

bremanandjk replied:
@agrawalravi90 I tried sampling server with 5000 number of multi-node jobs with each job requesting 100 chunks of 10 cpus. 

The main issue is observed only in single node job submission. It seems fine in multi-node jobs. 

Before fix: 
<img width="1645" alt="Before_fix_multinode" src="https://user-images.githubusercontent.com/17980660/88235077-1c232b80-cc48-11ea-9f1a-1ff5e2891c0e.png">

After fix: 

<img width="1645" alt="After_fix_multinode" src="https://user-images.githubusercontent.com/17980660/88235087-22190c80-cc48-11ea-98ed-39d4c62757d0.png">

agrawalravi90 replied:
A few questions:
- How many nodes did you have? how many ncpus on each node?
- I'm a bit confused why req_modifyjob() is taking so much time, especially in the first graph, did you not have enough cpus for ALL jobs to run?
- Can you please highlight free_nodes() in the graphs? You can search for a symbol in perf report just like you do in vim

bremanandjk replied:
* How many nodes did you have? how many ncpus on each node?
 In my setup, I have 500 vnodes and each node is having 100 cpus each. In total 50k cpus. 
* I'm a bit confused why req_modifyjob() is taking so much time, especially in the first graph, did you not have enough cpus for ALL jobs to run?
 The requested amount cpus are available in the test system
* Can you please highlight free_nodes() in the graphs? You can search for a symbol in perf report just like you do in vim
 I am not able to locate the free_nodes() while running in multi-node tests. I have tried various test setups, Its happening mostly in single node tests. Even searched in vim too.

bremanandjk replied:
Please ignore the previous results:

I have tried the setup suggested by @agrawalravi90  (Thanks) , In this setup, I could get the free_nodes symbol perfectly. 

Setup details: 

1) 50000 vnodes, each with 10 cpus. 
2) Submitted 5k jobs, each requesting 10 chunks of 10 ncpus,

Scanned the server in perf till the job completion with one scheduling cycle. 

Before fix:
<img width="1645" alt="master_multinode_50knodes" src="https://user-images.githubusercontent.com/17980660/88244627-5a2d4900-cc62-11ea-8cec-0dd1ff39a357.png">

After fix: 
<img width="1645" alt="master_after_fix_50knodes" src="https://user-images.githubusercontent.com/17980660/88244635-60232a00-cc62-11ea-9e5f-fd027769bd32.png">

bremanandjk commented:
@hirenvadalia I have renamed the function as you suggest. 

@subhasisb & @agrawalravi90 Thanks for your approvals, This PR is ready to merge.

hirenvadalia commented:
@bremanandjk I think you need to do force push again to resolve 2 Travis status issue. (This happens if you do back to back force push in PR)

bremanandjk commented:
Not sure about 2 Travis builds, But the Travis build is successful. One build is showing a green tick and another one not getting results back.

bremanandjk commented:
@hirenvadalia  I have already done force push, cant try it again, 

git push -f mine freenodes_master_pr
Everything up-to-date

bremanandjk commented:
Thanks, @hirenvadalia  for approval and merging.
